core/node/zk_os_tree_manager/src/batch.rs

```diff
@@ -0,0 +1,168 @@
+            writes.collect()
+        } else {
+            load_tree_writes_latency.observe();
+            // Otherwise, load writes' data from other tables.
+            Self::extract_logs_from_db(storage, l1_batch_number).await?
+        };
+
+        load_changes_latency.observe();
+        Ok(Some(Self { stats, tree_logs }))
+    }
+
+    async fn wait_for_tree_writes(
+        connection: &mut Connection<'_, Core>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Option<Vec<TreeWrite>>> {
+        const INTERVAL: Duration = Duration::from_millis(50);
+        const TIMEOUT: Duration = Duration::from_secs(5);
+
+        tokio::time::timeout(TIMEOUT, async {
+            loop {
+                if let Some(tree_writes) = connection
+                    .blocks_dal()
+                    .get_tree_writes(l1_batch_number)
+                    .await?
+                {
+                    break anyhow::Ok(tree_writes);
+                }
+                tokio::time::sleep(INTERVAL).await;
+            }
+        })
+        .await
+        .ok()
+        .transpose()
+    }
+
+    // Unlike with the Era tree, we don't need to filter out protective reads here because each L1 batch / VM run
+    // is a single block (i.e., a situation in which there are multiple storage logs in a batch for the same storage slot doesn't occur).
+    async fn extract_logs_from_db(
+        connection: &mut Connection<'_, Core>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Vec<TreeEntry>> {
+        let touched_slots_latency = METRICS.start_load_stage(LoadChangesStage::LoadTouchedSlots);
+        let touched_slots = connection
+            .storage_logs_dal()
+            .get_touched_slots_for_executed_l1_batch(l1_batch_number)
+            .await
+            .context("cannot fetch touched slots")?;
+        touched_slots_latency.observe_with_count(touched_slots.len());
+
+        let indices_latency = METRICS.start_load_stage(LoadChangesStage::LoadLeafIndices);
+        let hashed_keys_for_writes: Vec<_> =
+            touched_slots.keys().map(StorageKey::hashed_key).collect();
+        let l1_batches_for_initial_writes = connection
+            .storage_logs_dal()
+            .get_l1_batches_and_indices_for_initial_writes(&hashed_keys_for_writes)
+            .await
+            .context("cannot fetch initial writes batch numbers and indices")?;
+        indices_latency.observe_with_count(l1_batches_for_initial_writes.len());
+
+        // Sort tree logs by the enumeration index to get the correct order for inserts.
+        let mut tree_logs = BTreeMap::new();
+        for (storage_key, value) in touched_slots {
+            if let Some(&(initial_write_batch_for_key, leaf_index)) =
+                l1_batches_for_initial_writes.get(&storage_key.hashed_key())
+            {
+                // Filter out logs that correspond to deduplicated writes.
+                if initial_write_batch_for_key <= l1_batch_number {
```

Comment:
I think it can be simplified further: there will be no duplicated writes in `storage_logs` so filtering by `l1_batch_number` is needless. 
Off-topic: probably we won't even save `l1_batch_number` in `initital_writes` as it can be queried from `storage_logs` -- get the first miniblock number for given `hashed_key`. We still need the table to store hashed_key => leaf_index though. I'm not certain because we need to migrate old chains to zk os so we might still want to save `l1_batch_number`.

Reply:
If this doesn't produce much overhead, saving `l1_batch_number` could be beneficial at least for consistency.